In [3]:
import pandas as pd
import numpy as np

data=pd.read_csv('Assignment - Analytics Intern - Airlearn - sales_data_500.csv')
data.sample(30)

,Date,Order ID,Customer Name,Product Category,Product Name,Quantity Sold,Unit Price,Total Revenue,Payment Method,Delivery Status
190,2025-01-25,ORD10190,Edward Kim,Clothing,Sneakers,2,97.77,195.54,Credit Card,Delivered
175,2025-02-04,ORD10175,Diana Lee,Beauty,Lipstick,5,492.95,2464.75,Cash,Delivered
350,2024-12-14,ORD10350,Edward Kim,Electronics,Smartphone,4,62.28,249.12,Credit Card,Delivered
174,2025-01-31,ORD10174,Julia Adams,Electronics,Smartwatch,4,312.85,1251.40,Cash,Delivered
147,2024-10-08,ORD10147,Fiona Green,Clothing,Jeans,1,399.68,399.68,Credit Card,Pending
241,2024-10-06,ORD10241,Hannah White,Home Appliances,Air Fryer,5,293.13,1465.65,Credit Card,Delivered
331,2025-02-12,ORD10331,George Brown,Books,Non-Fiction,3,213.14,639.42,Cash,Delivered
467,2025-02-05,ORD10467,Bob Smith,Beauty,Lipstick,4,201.83,807.32,Credit Card,Pending
136,2024-10-23,ORD10136,Bob Smith,Home Appliances,Vacuum Cleaner,4,356.12,1424.48,Cash,Pending
91,2025-02-04,ORD10091,Fiona Green,Beauty,Foundation,2,361.73,723.46,Cash,Delivered


In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Date              500 non-null    object 
 1   Order ID          500 non-null    object 
 2   Customer Name     500 non-null    object 
 3   Product Category  500 non-null    object 
 4   Product Name      500 non-null    object 
 5   Quantity Sold     500 non-null    int64  
 6   Unit Price        500 non-null    float64
 7   Total Revenue     500 non-null    float64
 8   Payment Method    500 non-null    object 
 9   Delivery Status   500 non-null    object 
dtypes: float64(2), int64(1), object(7)
memory usage: 39.2+ KB


In [5]:
data['Date'] = pd.to_datetime(data['Date'])

In [6]:
data["Delivery Status"].unique()

array(['Delivered', 'Cancelled', 'Pending'], dtype=object)

Delivered: 1, Cancelled: 0, Pending: 2

In [7]:
data["Delivery Status"].replace({"Delivered": 1, "Cancelled": 0, "Pending": 2}, inplace=True)


/tmp/ipykernel_19830/2824775933.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data["Delivery Status"].replace({"Delivered": 1, "Cancelled": 0, "Pending": 2}, inplace=True)
/tmp/ipykernel_19830/2824775933.py:1: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  data["Delivery Status"].replace({"Deli

In [8]:
data["Payment Method"].unique()

array(['Cash', 'PayPal', 'Credit Card'], dtype=object)

{"Cash": 0, "PayPal": 1, "Credit Card": 2}

In [9]:
data["Payment Method"].replace({"Cash": 0, "PayPal": 1, "Credit Card": 2}, inplace=True)

/tmp/ipykernel_19830/204972507.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data["Payment Method"].replace({"Cash": 0, "PayPal": 1, "Credit Card": 2}, inplace=True)
/tmp/ipykernel_19830/204972507.py:1: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  data["Payment Method"].replace({"Cash": 0, "P

In [10]:
data["Product Category"].unique()

array(['Clothing', 'Books', 'Beauty', 'Electronics', 'Home Appliances'],
      dtype=object)

{"Clothing": 0, "Books": 1, "Beauty": 2, "Electronics": 3, "Home Appliances": 4}

In [11]:
data["Product Category"].replace({"Clothing": 0, "Books": 1, "Beauty": 2, "Electronics": 3, "Home Appliances": 4}, inplace=True)

/tmp/ipykernel_19830/2472627553.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data["Product Category"].replace({"Clothing": 0, "Books": 1, "Beauty": 2, "Electronics": 3, "Home Appliances": 4}, inplace=True)
/tmp/ipykernel_19830/2472627553.py:1: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  dat

Removed Order ID as its not important that much

In [12]:
data.drop(columns=['Order ID'], inplace=True)

In [13]:
data["Customer Name"].unique()

array(['Julia Adams', 'George Brown', 'Ian Black', 'Charlie Davis',
       'Hannah White', 'Bob Smith', 'Fiona Green', 'Alice Johnson',
       'Diana Lee', 'Edward Kim'], dtype=object)

In [14]:
name_mapping = {
    'Julia Adams': 0,
    'George Brown': 1,
    'Ian Black': 2,
    'Charlie Davis': 3,
    'Hannah White': 4,
    'Bob Smith': 5,
    'Fiona Green': 6,
    'Alice Johnson': 7,
    'Diana Lee': 8,
    'Edward Kim': 9
}

data["Customer Name"].replace(name_mapping, inplace=True)

/tmp/ipykernel_19830/2527371479.py:14: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data["Customer Name"].replace(name_mapping, inplace=True)
/tmp/ipykernel_19830/2527371479.py:14: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  data["Customer Name"].replace(name_mapping, inplace=True)


In [15]:
data.sample(30)

,Date,Customer Name,Product Category,Product Name,Quantity Sold,Unit Price,Total Revenue,Payment Method,Delivery Status
136,2024-10-23,5,4,Vacuum Cleaner,4,356.12,1424.48,0,2
132,2024-12-24,1,3,Laptop,4,33.65,134.60,0,1
273,2024-10-13,4,1,Non-Fiction,4,328.58,1314.32,1,1
227,2024-12-20,3,0,Jacket,3,159.76,479.28,0,1
210,2025-01-28,2,0,Sneakers,1,462.87,462.87,2,1
194,2025-02-07,8,2,Lipstick,5,64.38,321.90,2,2
342,2024-12-29,5,0,Sneakers,4,224.15,896.60,1,1
13,2024-11-28,7,1,Non-Fiction,2,364.34,728.68,0,1
82,2025-01-01,9,0,Sneakers,3,146.99,440.97,1,1
388,2024-09-21,0,4,Vacuum Cleaner,3,163.97,491.91,2,1


In [16]:
data.isnull().sum()

Date                0
Customer Name       0
Product Category    0
Product Name        0
Quantity Sold       0
Unit Price          0
Total Revenue       0
Payment Method      0
Delivery Status     0
dtype: int64

In [17]:
data.describe()

,Date,Customer Name,Product Category,Quantity Sold,Unit Price,Total Revenue,Payment Method,Delivery Status
count,500,500.000000,500.000000,500.00000,500.000000,500.000000,500.000000,500.00000
mean,2024-11-29 09:04:19.200000256,4.480000,1.974000,3.10600,258.501200,784.169120,0.962000,1.11000
min,2024-08-30 00:00:00,0.000000,0.000000,1.00000,10.190000,13.570000,0.000000,0.00000
25%,2024-10-14 00:00:00,2.000000,1.000000,2.00000,141.532500,333.405000,0.000000,1.00000
50%,2024-12-02 12:00:00,4.000000,2.000000,3.00000,260.955000,645.410000,1.000000,1.00000
75%,2025-01-12 00:00:00,7.000000,3.000000,4.00000,377.770000,1153.875000,2.000000,1.00000
max,2025-02-26 00:00:00,9.000000,4.000000,5.00000,496.560000,2467.200000,2.000000,2.00000
std,NaN,2.995253,1.332996,1.39379,140.097627,565.257614,0.798644,0.50439


Date
Count: 500 transactions.
Mean: November 29, 2024, 09:04:19—roughly the midpoint of the data collection period.
Min: August 30, 2024.
Max: February 26, 2025 (matches today’s date, suggesting the dataset is current up to now).
Quartiles:
25%: October 14, 2024.
50% (Median): December 2, 2024.
75%: January 12, 2025.
Insight: The data spans nearly six months, with a fairly symmetric distribution around late November/early December. The slight skew toward later dates (median closer to Q1 than Q3) suggests a possible uptick in transactions as we approach early 2025—perhaps holiday or seasonal effects.

Customer Name
Mean: 4.48 (between Hannah White [4] and Bob Smith [5]).
Min: 0 (Julia Adams).
Max: 9 (Edward Kim).
Std: 2.995—reasonable spread across 10 customers.
Quartiles: 2 (Ian Black), 4 (Hannah White), 7 (Alice Johnson).
Insight: With only 10 unique customers encoded 0–9, the mean suggests a slight tilt toward mid-to-higher-indexed customers like Hannah, Bob, and Alice. The standard deviation indicates good diversity—no single customer dominates excessively.

Product Category
Mean: 1.974 (~2, Beauty).
Min: 0 (Clothing).
Max: 4 (Home Appliances).
Std: 1.333—moderate variability.
Quartiles: 1 (Books), 2 (Beauty), 3 (Electronics).
Insight: Five categories (0–4: Clothing, Books, Beauty, Electronics, Home Appliances). The mean near 2 suggests Beauty and Electronics (2 and 3) are popular, with Books (1) also in play. Clothing (0) and Home Appliances (4) may be less frequent.

Quantity Sold
Mean: 3.106 units.
Min: 1, Max: 5.
Std: 1.394—fairly tight clustering.
Quartiles: 2, 3, 4.
Insight: Sales quantities range from 1 to 5, with a mean just above 3. Most transactions involve 2–4 units, suggesting small-to-medium batch purchases. The low standard deviation indicates consistency—few outliers.
Unit Price
Mean: $258.50.
Min: $10.19, Max: $496.56.
Std: $140.10—significant variation.
Quartiles: $141.53, $260.96, $377.77.
Insight: Prices vary widely, reflecting diverse products (e.g., cheap Comics vs. pricey Refrigerators). The mean near the median ($260.96) suggests a balanced distribution, but the high standard deviation flags heterogeneity—some items are budget, others premium.

Total Revenue
Mean: $784.17.
Min: $13.57, Max: $2,467.20.
Std: $565.26—large spread.
Quartiles: $333.41, $645.41, $1,153.88.
Insight: Revenue per transaction varies dramatically, driven by unit price and quantity. The mean exceeds the median, indicating a right skew—some high-value sales boost the average. This aligns with premium products like Vacuum Cleaners or Laptops in the sample.

Payment Method
Mean: 0.962 (~1, PayPal).
Min: 0 (Cash), Max: 2 (Credit Card).
Std: 0.799—decent spread.
Quartiles: 0 (Cash), 1 (PayPal), 2 (Credit Card).
Insight: Three methods (0: Cash, 1: PayPal, 2: Credit Card). The mean suggests PayPal is common, but all methods are used. The even quartile split confirms no single method dominates excessively.

Delivery Status
Mean: 1.11 (~1, Delivered).
Min: 0 (Cancelled), Max: 2 (Pending).
Std: 0.504—low variability.
Quartiles: 1, 1, 1—all Delivered.
Insight: Most orders are Delivered (1), with a mean just above 1 indicating some Pending (2) or Cancelled (0) statuses. The tight standard deviation and quartile sweep to 1 suggest high fulfillment reliability.

In [19]:
Q1 = data['Total Revenue'].quantile(0.25)
Q3 = data['Total Revenue'].quantile(0.75)
IQR = Q3 - Q1
df = data[~((data['Total Revenue'] < (Q1 - 1.5 * IQR)) | (data['Total Revenue'] > (Q3 + 1.5 * IQR)))]
df.head()

,Date,Customer Name,Product Category,Product Name,Quantity Sold,Unit Price,Total Revenue,Payment Method,Delivery Status
1,2024-12-19,1,1,Non-Fiction,4,246.75,987.00,1,1
2,2025-01-26,0,2,Perfume,5,336.56,1682.80,1,1
3,2024-11-02,2,0,Jacket,5,193.58,967.90,0,1
4,2024-10-03,3,2,Lipstick,5,282.25,1411.25,2,1
5,2024-11-20,4,3,Smartwatch,3,483.36,1450.08,2,1
